packages/eventual-send/src/index.js

```diff
@@ -71,33 +68,109 @@ export function makeHandledPromise() {
+        promiseToPendingHandler.delete(target);
         target = parent;
       }
     }
     return target;
   }
 
-  // This special handler accepts Promises, and forwards
-  // handled Promises to their corresponding fulfilledHandler.
+  /**
+   * This special handler accepts Promises, and forwards
+   * handled Promises to their corresponding fulfilledHandler.
+   *
+   * @type {Required<EHandler<any>>}
+   */
   let forwardingHandler;
   let handle;
 
-  function HandledPromise(executor, unsettledHandler = undefined) {
-    if (new.target === undefined) {
-      throw new Error('must be invoked with "new"');
+  /**
+   * @param {string} handlerName
+   * @param {EHandler<any>} handler
+   * @param {string} operation
+   * @param {any} o
+   * @param {any[]} opArgs
+   * @param {Promise<unknown>} [returnedP]
+   * @returns {any}
+   */
+  const dispatchToHandler = (
+    handlerName,
+    handler,
+    operation,
+    o,
+    opArgs,
+    returnedP,
+  ) => {
+    let actualOp = operation;
+    const triedOperations = [operation];
+    const isSendOnly = actualOp.endsWith('SendOnly');
+    const makeResult = result => (isSendOnly ? undefined : result);
+
+    if (isSendOnly) {
+      // We don't specify the resulting promise if it is sendonly.
+      returnedP = undefined;
+    }
+
+    if (isSendOnly && typeof handler[actualOp] !== 'function') {
+      // Substitute for sendonly with the corresponding non-sendonly operation.
+      actualOp = actualOp.slice(0, -'SendOnly'.length);
+      triedOperations.push(actualOp);
+    }
+
+    // Fast path: just call the actual operation.
+    if (typeof handler[actualOp] === 'function') {
+      return makeResult(handler[actualOp](o, ...opArgs, returnedP));
     }
+
+    if (actualOp === 'applyMethod') {
+      // Compose a missing applyMethod by get followed by applyFunction.
+      const [prop, args] = opArgs;
+      const getResultP = handle(o, 'get', [prop], undefined);
+      return makeResult(handle(getResultP, 'applyFunction', [args], returnedP));
+    }
+
+    if (actualOp === 'applyFunction') {
+      actualOp = 'applyMethod';
+      triedOperations.push(actualOp);
+      if (typeof handler[actualOp] === 'function') {
+        // Downlevel a missing applyFunction to applyMethod with undefined name.
+        const [args] = opArgs;
+        return makeResult(handler[actualOp](o, undefined, [args], returnedP));
+      }
+    }
+
+    const { details: X, quote: q } = assert;
+    assert.fail(
+      X`${q(handlerName)} is defined but has no methods needed for ${q(
+        operation,
+      )} (has ${q(Object.keys(handler).sort())})`,
+      TypeError,
+    );
+  };
+
+  /** @type {import('.').HandledPromiseConstructor} */
+  let HandledPromise;
+
+  /**
+   * @template R
+   * @param {import('.').HandledExecutor<R>} executor
+   * @param {EHandler<Promise<R>>} [pendingHandler]
+   * @returns {Promise<R>}
+   */
+  function HandledPromiseConstructor(executor, pendingHandler = undefined) {
```

Comment:
Confusing that `HandledPromiseConstructor` is both a type name for a function named `HandledPromise` and also a function name whose function type doesn't seem related to the `HandledPromiseConstructor` type.

Reply:
I'm at a loss for the correct TS magic to get it to recognise this function as a constructor that constructs `Promise<R>`s.  The parameters are correct, but TS *really* wants constructors to be defined with `class`.

Reply:
I tried the declaration you intended:

```suggestion
  /**
   * @type {HandledPromiseConstructor}
   */
  function HandledPromiseConstructor(executor, pendingHandler = undefined) {
```

to see what TS complains about. AFAICT, TS seems perfectly happy with this, both via `yarn lint` and vscode.

Reply:
Nevermind. Once I got to the end of index.js I saw the deeper dance going on. When I tried to reverse that to see what TS was complaining about, I saw plenty of complaints. So...

No change suggested. Resolving.